cdci_data_analysis/analysis/matrix_helper.py

```diff
@@ -0,0 +1,549 @@
+import time as time_
+import os
+import requests
+import glob
+import json
+import re
+import typing
+
+from ..analysis import tokenHelper
+from ..analysis.email_helper import humanize_age, humanize_future
+from ..analysis.exceptions import BadRequest, MissingRequestParameter
+from ..analysis.hash import make_hash
+from ..analysis.time_helper import validate_time
+from ..flask_app.sentry import sentry
+from ..app_logging import app_logging
+
+from datetime import datetime
+from jinja2 import Environment, FileSystemLoader
+from bs4 import BeautifulSoup
+from urllib import parse
+
+logger = app_logging.getLogger('matrix_helper')
+
+num_msgs_sending_max_tries = 5
+msg_sending_retry_sleep_s = .5
+
+class MatrixMessageNotSent(BadRequest):
+    pass
+
+
+class MultipleDoneMatrixMessage(BadRequest):
+    pass
+
+def timestamp2isot(timestamp_or_string: typing.Union[str, float]):
+    try:
+        timestamp_or_string = validate_time(timestamp_or_string).strftime("%Y-%m-%d %H:%M:%S")
+    except (ValueError, OverflowError, TypeError, OSError) as e:
+        logger.warning(f'Error when constructing the datetime object from the timestamp {timestamp_or_string}:\n{e}')
+        raise MatrixMessageNotSent(f"Matrix message not sent: {e}")
+
+    return timestamp_or_string
+
+
+def textify_matrix_message(html):
+    html = re.sub('<a href=(.*?)>(.*?)</a>', r'\2: \1', html)
+
+    soup = BeautifulSoup(html)
+
+    for elem in soup.find_all(["a", "p", "div", "h3", "br"]):
+        elem.replace_with(elem.text + "\n\n")
+
+    return soup.get_text()
+
+
+
+def get_first_submitted_matrix_message_time(scratch_dir):
```

Comment:
this duplicates largely what is done for email. You think it is feasible to make a common "notification" functionality which sends to matrix or email depending on some argument?

maybe it's too much effort, in this case let's leave it as it is. But let me know what you think.

Reply:
Yes, I thought about it, and I agree with you. I even considered implementing a 'plugin' system, where we expose an interface. Each time we integrate a new notification system, we create a new implementation of that interface.

In any case, for now, combining both Matrix and email notifications into a single functionality to avoid duplication as we have now, will require some effort, but it's not overly complex.

I chose this approach because we might want to modify the logic for Matrix notifications in the future. Currently, we're using the same logic as we do for emails. But considering that Matrix is primarily used by us, we might want more customized settings tailored to our needs (more simple, more complex?) eventually in the future.

Anyway, it is good to keep in mind that it could be a refactoring and improvement for the dispatcher